api/v1alpha1/mirrorpeer_webhook.go

```diff
@@ -0,0 +1,95 @@
+/*
+Copyright 2021 Red Hat OpenShift Data Foundation.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"fmt"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+// log is for logging in this package.
+var mirrorpeerlog = logf.Log.WithName("mirrorpeer-resource")
+
+func (r *MirrorPeer) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(r).
+		Complete()
+}
+
+//+kubebuilder:webhook:path=/mutate-multicluster-odf-openshift-io-v1alpha1-mirrorpeer,mutating=true,failurePolicy=fail,sideEffects=None,groups=multicluster.odf.openshift.io,resources=mirrorpeers,verbs=create;update,versions=v1alpha1,name=mmirrorpeer.kb.io,admissionReviewVersions=v1
+
+var _ webhook.Defaulter = &MirrorPeer{}
+
+// Default implements webhook.Defaulter so a webhook will be registered for the type
+func (r *MirrorPeer) Default() {
+	mirrorpeerlog.Info("default", "name", r.Name)
+
+	if len(r.Spec.SchedulingIntervals) == 0 {
+		r.Spec.SchedulingIntervals = []string{"5m"}
+	}
+}
+
+//+kubebuilder:webhook:path=/validate-multicluster-odf-openshift-io-v1alpha1-mirrorpeer,mutating=false,failurePolicy=fail,sideEffects=None,groups=multicluster.odf.openshift.io,resources=mirrorpeers,verbs=create;update;delete,versions=v1alpha1,name=vmirrorpeer.kb.io,admissionReviewVersions=v1alpha1
+
+var _ webhook.Validator = &MirrorPeer{}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type
+func (r *MirrorPeer) ValidateCreate() error {
+	mirrorpeerlog.Info("validate create", "name", r.Name)
+
+	return r.ValidateMirrorPeer(r)
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type
+func (r *MirrorPeer) ValidateUpdate(old runtime.Object) error {
+	mirrorpeerlog.Info("validate update", "name", r.Name)
+	oldMirrorPeer, ok := old.(*MirrorPeer)
+	if !ok {
+		return fmt.Errorf("error casting old object to MirrorPeer")
+	}
+
+	for idx, pr := range oldMirrorPeer.Spec.Items {
+		if pr.StorageClusterRef != r.Spec.Items[idx].StorageClusterRef || pr.ClusterName != r.Spec.Items[idx].ClusterName {
+			return fmt.Errorf("cannot update MirrorPeer: storageclusterref or clustername cannot be changed")
+		}
+	}
```

Comment:
Items position can change between updates.
We should check that new items are not added and old items are not removed.

Reply:
Changed the logic